### PR TITLE
PaddleScience 论文复现 Extended Physics-Informed Neural Networks (XPINNs)

### PR DESCRIPTION
**Pull request by co63oc**
### PR types

Others
### PR changes

Others
### Describe

https://github.com/PaddlePaddle/Paddle/issues/55663#paddlepaddle14

赛题十四：PaddleScience 领域经典论文复现
论文一
题目：Extended Physics-Informed Neural Networks (XPINNs)

网络结构已转换一致，修改为  self.optimizer.clear_grad() 可以对齐精度

```
            loss = loss1_value + loss2_value + loss3_value
            loss.backward()
            self.optimizer.step()
            self.optimizer.clear_grad()
```

Test Loss 信息，迭代次数修改为501，tensorflow 1.13.2-gpu 版本，Loss 差异
| Test Loss | tensorflow | paddle |
| - | - | - |
| | 1.794519e-02 | 2.138322e-01 |

tensorflow运行
图片

paddle运行
图片

### CODE REVIEW

File: jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+        MSE_history3 = []
+        l2_err2 = []
+        l2_err3 = []
+
+        for it in range(nIter):
+            loss1_value, loss2_value, loss3_value = self.model(self.dataset)
+            loss = loss1_value + loss2_value + loss3_value
+            loss.backward()
+            self.optimizer.step()
+            loss.clear_grad()
```
Comment (HydrogenSulfate) [inline jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py:306]: 看到精度没对齐，应该是这行的问题？`Tensor.clear_grad` 只能清空 `Tensor` 自己的梯度，梯度更新完应该调用`self.optimizer.clear_grad()`

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py (hunk 1 above)
Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 修改为 self.optimizer.clear_grad() 可以对齐精度

File: jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py [hunk 2]
```diff
@@ -352,7 +352,7 @@ def predict(self, X_star1, X_star2, X_star3):
     layers2 = [2, 20, 20, 20, 20, 1]
     layers3 = [2, 25, 25, 25, 1]
 
-    Max_iter = 501
+    Max_iter = 701
```
Comment (HydrogenSulfate) [inline jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py:355]: 源代码好像是501，这里为改成701的原因是什么呢

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py:355]: 是用来对齐精度，迭代501次Loss较大

Comment (HydrogenSulfate) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py:355]: > 是用来对齐精度，迭代501次Loss较大

一般论文复现默认使用相同的训练策略，如果转换成paddle后，需要增大epoch来对齐精度，可能是某些地方仍然没有对齐，比如参数初始化之类的细节

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py:355]: 修改tensorflow 版本1.13.2-gpu，paddle 精度可以对齐

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: > > 修改tensorflow 版本1.13.2-gpu，paddle 精度可以对齐，PR描述已更新
> 
> OK，我运行下试试看

不是tensorflow 版本问题，是代码有改动，我再看下

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 图片
图片

使用 paddle.nn.initializer.TruncatedNormal 提示不支持 float64，是支持float32

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 参数有W, b, a，b和a是固定初始值, W使用 XavierNormal和使用 TruncatedNormal  float32精度不对齐

Comment (HydrogenSulfate) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: > 参数有W, b, a，b和a是固定初始值, W使用 XavierNormal和使用 TruncatedNormal float32精度不对齐

PaddleScience里有这个API，看下能不能等价复现一下，https://github.com/PaddlePaddle/PaddleScience/blob/develop/ppsci/utils/initializer.py#L151

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 图片
同样错误不支持float64

Comment (HydrogenSulfate) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: > 图片 同样错误不支持float64

看了一下，是这部分代码有点问题，可以再to_tensor末尾加上, _tensor.dtype

``` py
_tensor = paddle.multiply(_tensor, paddle.to_tensor(std * math.sqrt(2.0), _tensor.dtype))
_tensor = paddle.add(_tensor, paddle.to_tensor(mean, _tensor.dtype))
```

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 图片
结果为0

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]:
```
    def xavier_init(self, size):
        in_dim = size[0]
        out_dim = size[1]        
        import ppsci
        param = paddle.empty(size, "float64")
        param = ppsci.utils.initializer.trunc_normal_(param, 0.0, 1.0)
        return paddle.nn.initializer.Assign(param)
```

paddle.nn.initializer.Assign 不支持float64
图片

Comment (HydrogenSulfate) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: > 

暂时用lambda代替下看看
``` py
import numpy as np

x_np = np.zeros([2, 2], "float64")

w = paddle.create_parameter(
    x_np.shape,
    dtype="float64",
    default_initializer=lambda p_ten, _: p_ten.set_value(x_np)
)

print(w)

"""
Tensor(shape=[2, 2], dtype=float64, place=Place(gpu:0), stop_gradient=False,
       [[0., 0.],
        [0., 0.]])
"""
```

Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 修改后精度为 4.684577e-02, tensorflow精度是 1.794519e-02
```
    def w_init(self, size):
        in_dim = size[0]
        out_dim = size[1]        
        xavier_stddev = np.sqrt(2/(in_dim + out_dim))
        param = paddle.empty(size, "float64")
        param = ppsci.utils.initializer.trunc_normal_(param, 0.0, xavier_stddev)
        return lambda p_ten, _: p_ten.set_value(param)
```

图片

Comment (HydrogenSulfate) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: > 修改后精度为 4.684577e-02, tensorflow精度是 1.794519e-02
> 
> ```
>     def w_init(self, size):
>         in_dim = size[0]
>         out_dim = size[1]        
>         xavier_stddev = np.sqrt(2/(in_dim + out_dim))
>         param = paddle.empty(size, "float64")
>         param = ppsci.utils.initializer.trunc_normal_(param, 0.0, xavier_stddev)
>         return lambda p_ten, _: p_ten.set_value(param)
> ```
> 
> 图片

分布相同的情况下，两个框架生成的随机数也是存在不同的，这个diff大概率是合理的，我这边明天换个随机种子再跑下paddle看看

File: jointContribution/XPINNs/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+       }
+
+Recommended software versions: TensorFlow 1.14, Python 3.6, Latex (for plotting figures)
+
+For any queries regarding the XPINN code, feel free to contact me : <ameya_jagtap@brown.edu>, <ameyadjagtap@gmail.com>
+
+--------------------
+原仓库 <https://github.com/AmeyaJagtap/XPINNs.git>
+
+安装latex软件包，python依赖
```
Comment (HydrogenSulfate) [inline jointContribution/XPINNs/README.md:91]: 详细说明下latex软件包怎么安装的，我这边pip 安装latex仍然会报错

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: jointContribution/XPINNs/README.md (hunk 3 above)
Comment (co63oc) [reply jointContribution/XPINNs/README.md]: 下面有代码
图片

File: jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+            A.append(a)
+        return weights, biases, A
+
+    def w_init(self, size):
+        in_dim = size[0]
+        out_dim = size[1]
+        xavier_stddev = np.sqrt(2 / (in_dim + out_dim))
+        param = paddle.empty(size, "float64")
+        param = ppsci.utils.initializer.trunc_normal_(param, 0.0, xavier_stddev)
+        return lambda p_ten, _: p_ten.set_value(param)
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 1

File: jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+                default_initializer=paddle.nn.initializer.Constant(0.05),
+            )
+
+            self.add_parameter(name_prefix + "_W_" + str(l), W)
+            self.add_parameter(name_prefix + "_b_" + str(l), b)
+            self.add_parameter(name_prefix + "_a_" + str(l), a)
+            weights.append(W)
+            biases.append(b)
+            A.append(a)
+        return weights, biases, A
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/XPINNs/XPINN_2D_PoissonsEqn.py]: 修改为 amplitudes

File: jointContribution/XPINNs/.gitignore
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/XPINNs/.gitignore:1]: 1

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
